File: 0155-min-stack/0155-min-stack.cs
public class MinStack {

    Stack<int> stack = null;
    Stack<int> minimumStack = null;
    
    public MinStack() {
        stack = new Stack<int>();
        minimumStack = new Stack<int>();
    }
    
    public void Push(int val) {
        
        if(minimumStack.Count > 0)
        {
            if(minimumStack.Peek() >= val)
            {
                minimumStack.Push(val);
            }
        }
        else
        {
            minimumStack.Push(val);
        }
        
        stack.Push(val);
        
    }
    
    public void Pop() {
        
        var temp = stack.Pop();
        if(minimumStack.Count > 0 && temp == minimumStack.Peek())
        {
            minimumStack.Pop();
        }
        
    }
    
    public int Top() {
        return stack.Peek();
    }
    
    public int GetMin() {
        return minimumStack.Peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.Push(val);
 * obj.Pop();
 * int param_3 = obj.Top();
 * int param_4 = obj.GetMin();
 */

File: 0155-min-stack/0155-min-stack.cpp
class MinStack {
public:
    vector<pair<int, int>> v;
    int minm;
    MinStack() {
        minm = INT_MAX;
    }
    
    void push(int val) {

        minm = v.empty() ? val : min(val, v.back().second);
        v.push_back({val, minm});
        // if(val < minm) 
        //     minm = val;
        // v.push_back(pair<int, int>(val, minm));
        
    }
    
    void pop() {
        v.pop_back();
    }
    
    int top() {
        return v[v.size()-1].first;
    }
    
    int getMin() {
        return v[v.size()-1].second;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */